src/api/schema/log_events/mod.rs

```diff
@@ -0,0 +1,170 @@
+        let mut interval = time::interval(time::Duration::from_millis(interval));
+
+        // Collect a vector of results, with a capacity of `limit`. As new `LogEvent`s come in,
+        // they will be sampled and added to results.
+        let mut results = Vec::<LogEventPayload>::with_capacity(limit);
+
+        // Random number generator to allow for sampling. Speed trumps cryptographic security here.
+        // The RNG must be Send + Sync to use with the `select!` loop below.
+        let mut rng = SmallRng::from_entropy();
+
+        // Keep a count of the batch size, which will be used as a seed for random eviction
+        // per the sampling strategy used below.
+        let mut batch = 0;
+
+        loop {
+            select! {
+                // Process `TapPayload`s. A tap payload could contain a `LogEvent` or a notification.
+                // Notifications are emitted immediately; log events buffer until the next `interval`.
+                Some(tap) = tap_rx.next() => {
+                    let tap = tap.into();
+
+                    // Emit notifications immediately; these don't count as a 'batch'.
+                    if let LogEventPayload::Notification(_) = tap {
+                        // If an error occurs when sending, the subscription has likely gone
+                        // away. Break the loop to terminate the thread.
+                        if log_tx.send(vec![tap]).await.is_err() {
+                            break;
+                        }
+                    } else {
+                        // A simple implementation of "Algorithm R" per
+                        // https://en.wikipedia.org/wiki/Reservoir_sampling. As we're unable to
+                        // pluck the nth result, this is chosen over the more optimal "Algorithm L"
+                        // since discarding results isn't an option.
+                        if limit > results.len() {
+                            results.push(tap);
+                        } else {
+                            let random_number = rng.gen_range(0..batch);
+                            if random_number < results.len() {
+                                results[random_number] = tap;
+                            }
+                        }
+                        // Increment the batch count, to be used for the next Algo R loop
+                        batch += 1;
+                    }
+                }
+                _ = interval.tick() => {
+                    // If there are any existing results after the interval tick, emit.
+                    if !results.is_empty() {
+                        // Reset the batch count, to adjust sampling probability for the next round.
+                        batch = 0;
+
+                        // Since events will appear out of order per the random sampling
+                        // strategy, drain the existing results and sort by timestamp.
```

Comment:
Notably this could reorder events that have the same timestamp. I'm curious what you think about that. At the least we should document this behavior as I could see a user being confused.

Reply:
AFAICT, sort appears to be stable to insertion order into the original vector, when comparing keys:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=8b88c8f986d67b5270bd7784a677bbc2

So I don't _think_ this should arbitrarily reshuffle timestamps of matching events; they should be emitted in the order that `tap_rx.next()` receives them.

Reply:
That's true, the sorting is stable, but the insertion order here is random after the buffer is filled and it randomly chooses elements to eject.

Reply:
Ah yes, that's true. If sorting by timestamps is likely to be surprising, we could go another way, and give each log an incrementing key that's sorted by instead?

Reply:
If it's not too much effort, I think that would be better. That's the sort of subtle behavior that I could see leading to confusion in rare circumstances.

Reply:
Added sorting by batch key in 7285387bafbb3c3cec360ffc1f3d1e4ca01e5882.

This is using `slice::sort_by_key`, which is O(m * n * log(n)) worst-case. There may be some easy wins moving to a method from [this table](https://crates.io/crates/sorting_rs).

Since the max input limit is currently 10,000, there's also https://crates.io/crates/count_sort which appear to have interesting performance up to `i16`, but it's only implemented for value sorting and uses unsafe.

Just an FYI for potential future optimisation work.